internal/util/importutilv2/binlog/reader.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type reader struct {
+	cm     storage.ChunkManager
+	schema *schemapb.CollectionSchema
+
+	deleteData *storage.DeleteData
+	insertLogs map[int64][]string // fieldID -> binlogs
+
+	readIdx int
+	filters []Filter
+}
+
+func NewReader(cm storage.ChunkManager,
+	schema *schemapb.CollectionSchema,
+	paths []string,
+	tsStart,
+	tsEnd uint64,
+) (*reader, error) {
+	r := &reader{
+		cm:     cm,
+		schema: schema,
+	}
+	err := r.Init(paths, tsStart, tsEnd)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (r *reader) Init(paths []string, tsStart, tsEnd uint64) error {
+	if tsStart != 0 || tsEnd != math.MaxUint64 {
+		r.filters = append(r.filters, FilterWithTimeRange(tsStart, tsEnd))
+	}
+	insertLogs, deltaLogs, err := listBinlogs(r.cm, paths)
+	if err != nil {
+		return err
+	}
+	err = verify(r.schema, insertLogs)
+	if err != nil {
+		return err
+	}
+	r.insertLogs = insertLogs
+	if len(deltaLogs) > 0 {
+		r.deleteData, err = r.ReadDelete(deltaLogs, tsStart, tsEnd)
+		if err != nil {
+			return err
+		}
+		deleteFilter, err := FilterWithDelete(r)
+		if err != nil {
+			return err
+		}
+		r.filters = append(r.filters, deleteFilter)
+	}
+	return nil
+}
+
+func (r *reader) ReadDelete(deltaLogs []string, tsStart, tsEnd uint64) (*storage.DeleteData, error) {
+	deleteData := storage.NewDeleteData(nil, nil)
+	for _, path := range deltaLogs {
+		reader, err := newBinlogReader(r.cm, path)
+		if err != nil {
+			return nil, err
+		}
+		rowsSet, err := readData(reader, storage.DeleteEventType)
+		if err != nil {
+			return nil, err
+		}
+		for _, rows := range rowsSet {
+			for _, row := range rows.([]string) {
+				dl := &storage.DeleteLog{}
+				err = json.Unmarshal([]byte(row), dl)
+				if err != nil {
+					return nil, err
+				}
+				if dl.Ts >= tsStart && dl.Ts <= tsEnd {
+					deleteData.Append(dl.Pk, dl.Ts)
+				}
+			}
+		}
+	}
+	return deleteData, nil
+}
+
+func (r *reader) Next(count int64) (*storage.InsertData, error) {
+	insertData, err := storage.NewInsertData(r.schema)
+	if err != nil {
+		return nil, err
+	}
+	if r.readIdx == len(r.insertLogs[0]) {
+		return nil, nil
+	}
+	for fieldID, binlogs := range r.insertLogs {
+		field := typeutil.GetField(r.schema, fieldID)
+		if field == nil {
+			return nil, merr.WrapErrFieldNotFound(fieldID)
+		}
+		path := binlogs[r.readIdx]
+		cr, err := newColumnReader(r.cm, field, path)
+		if err != nil {
+			return nil, err
+		}
+		fieldData, err := cr.Next(count)
+		if err != nil {
+			cr.Close()
+			return nil, err
+		}
+		cr.Close()
+		insertData.Data[field.GetFieldID()] = fieldData
+	}
+	insertData, err = r.Filter(insertData)
+	if err != nil {
+		return nil, err
+	}
+	r.readIdx++
+	return insertData, nil
+}
+
+func (r *reader) Filter(insertData *storage.InsertData) (*storage.InsertData, error) {
```

Comment:
Add a step first attempts to determine if filtering will occur. If it is determined that filtering will not take place, the function can directly return the previously obtained insertData, thus avoiding unnecessary copying.